backend/windmill-api/src/jobs.rs

```diff
@@ -195,1826 +251,1217 @@ impl RunJobQuery {
+        let status =
+            serde_json::from_value::<FlowStatus>(flow.flow_status.context("no flow status")?)
+                .context("deserialize flow status")?;
+        if matches!(status.current_step(), Some(FlowStatusModule::WaitingForEvents { job, .. }) if job == &job_id)
+        {
+            sqlx::query!(
+                "UPDATE queue SET suspend = $1 WHERE id = $2",
+                suspend,
+                flow.id,
+            )
+            .execute(&mut tx)
+            .await?;
+        }
+    }
+
+    tx.commit().await?;
+    Ok(StatusCode::CREATED)
 }
 
-pub async fn run_wait_result_job_by_hash(
-    authed: Authed,
-    Extension(user_db): Extension<UserDB>,
-    Path((w_id, script_hash)): Path<(String, ScriptHash)>,
-    axum::Json(args): axum::Json<Option<Map<String, Value>>>,
-    Query(run_query): Query<RunJobQuery>,
-) -> error::JsonResult<serde_json::Value> {
-    let hash = script_hash.0;
-    let mut tx = user_db.clone().begin(&authed).await?;
-    let path = get_path_for_hash(&mut tx, &w_id, hash).await?;
-    let scheduled_for = run_query.get_scheduled_for(&mut tx).await?;
+pub async fn cancel_suspended_job(
+    /* unauthed */
+    Extension(db): Extension<DB>,
+    Path((w_id, job, resume_id, secret)): Path<(String, Uuid, u32, String)>,
+    Query(approver): Query<QueryApprover>,
+) -> error::Result<String> {
+    let mut tx = db.begin().await?;
+    let key = get_workspace_key(&w_id, &mut tx).await?;
+    let mut mac = HmacSha256::new_from_slice(key.as_bytes()).map_err(to_anyhow)?;
+    mac.update(job.as_bytes());
+    mac.update(resume_id.to_be_bytes().as_ref());
+    if let Some(approver) = approver.approver.clone() {
+        mac.update(approver.as_bytes());
+    }
+    mac.verify_slice(hex::decode(secret)?.as_ref())
+        .map_err(|_| anyhow::anyhow!("Invalid signature"))?;
 
-    let (uuid, tx) = push(
-        tx,
+    let whom = approver.approver.unwrap_or_else(|| "unknown".to_string());
+    let parent_flow = get_root_job(db, &w_id, job).await?;
+    let (mut tx, job) = windmill_queue::cancel_job(
+        &whom,
+        Some("approval request disapproved".to_string()),
+        parent_flow,
         &w_id,
-        JobPayload::ScriptHash { hash: ScriptHash(hash), path },
-        args,
-        &authed.username,
-        owner_to_token_owner(&authed.username, false),
-        scheduled_for,
-        None,
-        run_query.parent_job,
-        false,
-        false,
+        tx,
     )
     .await?;
+    if job.is_some() {
+        audit_log(
+            &mut tx,
+            &whom,
+            "jobs.disapproval",
+            ActionKind::Delete,
+            &w_id,
+            Some(&parent_flow.to_string()),
+            None,
+        )
+        .await?;
+    }
     tx.commit().await?;
-
-    run_wait_result(authed, Extension(user_db), uuid, Path((w_id, script_hash))).await
+    Ok("Flow of job cancelled".to_string())
 }
```

Comment:
This had complicated merge changes, due to the approval changes. I think I did it right, but taking a look is probably a good idea.